velox/exec/tests/QueryTraceTest.cpp

```diff
@@ -0,0 +1,222 @@
+    connector::hive::LocationHandle::registerSerDe();
+    connector::hive::HiveColumnHandle::registerSerDe();
+    connector::hive::HiveInsertTableHandle::registerSerDe();
+    core::PlanNode::registerSerDe();
+    core::ITypedExpr::registerSerDe();
+    registerPartitionFunctionSerDe();
+  }
+
+ protected:
+  void SetUp() override {
+    if (!isRegisteredVectorSerde()) {
+      serializer::presto::PrestoVectorSerde::registerVectorSerde();
+    }
+    filesystems::registerLocalFileSystem();
+    HiveConnectorTestBase::SetUp();
+  }
+
+  static VectorFuzzer::Options getFuzzerOptions() {
+    return VectorFuzzer::Options{
+        .vectorSize = 16,
+        .nullRatio = 0.2,
+        .stringLength = 1024,
+        .stringVariableLength = false,
+        .allowLazyVector = false,
+    };
+  }
+
+  QueryTracerTest() : vectorFuzzer_{getFuzzerOptions(), pool_.get()} {
+    filesystems::registerLocalFileSystem();
+  }
+
+  RowTypePtr generateTypes(size_t numColumns) {
+    std::vector<std::string> names;
+    names.reserve(numColumns);
+    std::vector<TypePtr> types;
+    types.reserve(numColumns);
+    for (auto i = 0; i < numColumns; ++i) {
+      names.push_back(fmt::format("c{}", i));
+      types.push_back(vectorFuzzer_.randType((2)));
+    }
+    return ROW(std::move(names), std::move(types));
+    ;
+  }
+
+  static bool checkPlan(
+      const core::PlanNodePtr& left,
+      const core::PlanNodePtr& right) {
+    return (left->id() == right->id()) && (left->name() == right->name());
+  }
+
+  bool isSamePlan(
+      const core::PlanNodePtr& left,
+      const core::PlanNodePtr& right) {
+    if (!checkPlan(left, right)) {
+      return false;
+    }
+
+    if (left->sources().size() != right->sources().size()) {
+      return false;
+    }
+
+    for (auto i = 0; i < left->sources().size(); ++i) {
+      isSamePlan(left->sources().at(i), right->sources().at(i));
+    }
+    return true;
+  }
+
+  VectorFuzzer vectorFuzzer_;
+};
+
+TEST_F(QueryTracerTest, traceData) {
+  const auto rowType = generateTypes(5);
+  std::vector<RowVectorPtr> rows;
+  constexpr auto numBatch = 3;
+  rows.reserve(numBatch);
+  for (auto i = 0; i < numBatch; ++i) {
+    rows.push_back(vectorFuzzer_.fuzzInputRow(rowType));
+  }
+
+  const auto outputDir = TempDirectoryPath::create();
+  auto writer = trace::QueryDataWriter(outputDir->getPath());
+  for (auto i = 0; i < numBatch; ++i) {
+    writer.write(rows[i]);
+  }
+  writer.finish();
+
+  const auto reader = trace::QueryDataReader(outputDir->getPath());
+  RowVectorPtr actual;
+  reader.read(actual);
+  RowVectorPtr tmp;
+  while (reader.read(tmp)) {
+    actual->append(tmp.get());
+  }
+  RowVectorPtr expected = rows[0];
+  for (int i = 1; i < numBatch; ++i) {
+    expected->append(rows[i].get());
+  }
+  const auto sz = actual->size();
+  ASSERT_EQ(sz, expected->size());
+  for (auto i = 0; i < sz; ++i) {
+    actual->compare(expected.get(), i, i, {.nullsFirst = true});
+  }
```

Comment:
Have you verified the summary file?

Reply:
Yes, it is verified in the QueryDataReader's constructor, the `getTraceDataType()` read the summary file to extract the data type.